Comments: add chronological sorting for comments list

vindl opened the pull request:
Resolves https://github.com/Automattic/wp-calypso/issues/14118

Adds the ability to chronologically sort the list of comments by selecting
either the 'Newest first' or 'Oldest first' option. The 'Newest first' will be
used as default.

#### Visual changes

![peek 2017-09-22 13-02](https://user-images.githubusercontent.com/1182160/30741802-a37cce88-9f96-11e7-9cd1-2bc1be938c30.gif)

#### Testing instructions

1. Navigate to `comments/approved/{siteUrl}`.
2. Verify that chronological order of comments is respecting the selected setting.
3. Test on sites with a larger number of comments and verify that ordering persists after pagination.
4. After switching sort order the page should reset to first one.
5. Test with Jetpack sites.

vindl commented on client/my-sites/comments/constants.js:
```diff
@@ -0,0 +1,8 @@
+// Maximum number of comments that will be shown per page.
```
I considered moving this file to `client/lib/comments`, but not sure if it's worth it to create a new folder there just for this purpose.

Copons replied:
I'm personally opposed to it. These constants are only relevant to Comment Management, better not clutter `lib` with something that might confuse other devs.

Copons commented:
Tested and it looks and works great!

Before diving into the code, I've got two nitpicks:

- I'd decrease the sort buttons right margin to `16px`, to reduce the chances of elements overlapping on small-ish screens.

- I've noticed that changing sorting doesn't affect the pagination.
This is something I've always wondered myself when working on paginated and sorted lists.
  - Should changing sorting send back to the first page?
  - Should it stays on the current page? But then, the results in the current page might not be relevant anymore.
  - Should only change the order of the results shown in the current page? (Likely too complicated to realize, and too weird UX.)

I'm probably more into option 1, but it's definitely worth discussing.

vindl commented:
> I'd decrease the sort buttons right margin to 16px, to reduce the chances of elements overlapping on small-ish screens.

@Copons No objections to this, I'll just wait for design review before changing it.

> I've noticed that changing sorting doesn't affect the pagination.
This is something I've always wondered myself when working on paginated and sorted lists.

Good point! I'm leaning towards option 1 too, especially since paging info is usually not visible on the screen before scrolling (did we consider moving/duplicating it at the top?), so the user might not be looking at the newest/oldest comment after clicking the sorting option. Let's also wait for other folks to pitch in on this.

folletto commented:
I'm not really understanding why someone would sort by oldest, but... as it was the request, I assume there was a good reason. :)

Design wise, note that for these selector we shouldn't be using buttons, but the [SegmentedControl](https://wpcalypso.wordpress.com/devdocs/design/segmented-control):

![screen shot 2017-09-22 at 12 41 10](https://user-images.githubusercontent.com/4389/30742927-633673f4-9f93-11e7-9399-e40c0924a24f.png)

I know the difference is minor, but it's the semantic of the component so if we update it in the future it will behave properly.

> I've noticed that changing sorting doesn't affect the pagination.

This is a big issue. I'd agree to start with option 1 because it's simpler. 

The only other alternative I think makes sense is to keep the page "as if" the person started with the new ordering, so if switching on page 3 from Newest to Oldest, it should show page 3 of Oldest. But... let's start with going back to page 1. ;)

vindl commented:
> I'm not really understanding why someone would sort by oldest, but... as it was the request, I assume there was a good reason. :)

@folletto My guess is that it might useful for users that are going through a long list of pending comments and prefer to review them on first-come first-served basis.

> Design wise, note that for these selector we shouldn't be using buttons, but the SegmentedControl.

Yes, makes much more sense to use these, thanks! :)

folletto commented:
> My guess is that it might useful for users that are going through a long list of pending comments and prefer to review them on first-come first-served basis.

Ah yes, right, in the pending view one would see ONLY pending. Now I get it, thanks!

vindl commented:
@folletto and @Copons, I've made some updates in 2f6fe8d719026b9c23ee341fa08ad4a3cc9eb71f according to your feedback:

- decreased right margin of sort buttons to `16px`;
- added switching to first page after selecting sort order;
- refactored code to use `SegmentedControl` component instead of buttons.

folletto commented:
Design wise 👍

Copons commented on client/my-sites/comments/comment-list/index.jsx:
```diff
@@ -90,16 +94,29 @@ export class CommentList extends Component {
+				// Default to showing newest comments first.
+				return comments.sort( ( a, b ) => b - a );
+		}
+	};
```
We could simplify this with `_.orderBy`:

```js
const { sortOrder } = this.state;
return orderBy( comments, null, OLDEST_FIRST === sortOrder ? 'asc' : 'desc' );
```

vindl replied:
I was tinkering with that approach too, but left it like this in case we wanted to add other sorting methods in the future. On the other hand, since we can only compare ids here, additional sorting methods will probably require a more involved refactor, and at that point we'll probably have to rewrite most of it anyway - so I'm going to simplify it as you've suggested. :)

Actually, I was already prepared for this and _wisely_ used `asc` and `desc` for corresponding constant names, so we can simplify it even further to:
```javascript
return orderBy( comments, null, this.state.sortOrder );
```
:smile: 

![choosewisely](https://user-images.githubusercontent.com/1182160/30753202-e584ca6a-9fbe-11e7-8063-a0b445c2881a.jpg)

Copons replied:
![choosewisely](https://media.giphy.com/media/ZgYBhq1x7L1bW/200.gif)

Copons commented:
Other than that one minor comment, this looks definitely good to me.
I've also tested it in `rtl` and it works flawlessly.

🚢 whenever you feel comfy with it!

megsfulton commented:
Two quick thoughts: 

1) Could we use the styling of the segmented controller in the first row instead of the primary? I'm finding the blue in the toggle pulls a bit more attention than it probably deserves for a sort control 
<img width="425" alt="screen shot 2017-09-22 at 11 11 19 am" src="https://user-images.githubusercontent.com/161697/30758428-37e74fb8-9f87-11e7-844c-fe566160815d.png">

2. Maybe not a good time for this feedback, but did we consider a dropdown selector instead of a segmented control for the Newest/Oldest sort? It makes for one less button-y thing in the header menu and is a little bit less wide for small screens. The segmented controller works though so maybe we just roll with it for now.

<img width="833" alt="screen shot 2017-09-22 at 11 27 34 am" src="https://user-images.githubusercontent.com/161697/30758905-0f20fc9e-9f89-11e7-948c-8c8f77e620da.png">

vindl commented:
> Could we use the styling of the segmented controller in the first row instead of the primary? I'm finding the blue in the toggle pulls a bit more attention than it probably deserves for a sort control

Sure, I agree that blue color for this control makes it more pronounced than it needs to be.

> Maybe not a good time for this feedback, but did we consider a dropdown selector instead of a segmented control for the Newest/Oldest sort? It makes for one less button-y thing in the header menu and is a little bit less wide for small screens. The segmented controller works though so maybe we just roll with it for now.

If we end up adding more sort options I think this would be a good change, but at present state I'd like to avoid introducing another click to get to it. Also, on mobile screens it might look weird when both dropdowns are expanded.

gwwar commented on client/my-sites/comments/style.scss:
```diff
@@ -68,6 +68,11 @@
 
+.comment-navigation__sort-buttons {
+	margin-right: 16px;
+	float: left;
```
Were we unable to use flexbox?

vindl replied:
No, we can use flexbox too to achieve the same result. Any particular reason to prefer one over the other in this case?

Copons replied:
If you change to flex, please check if `rtl` works as expected.
AFAIK, floats are reversed automagically, while I'm not sure it happens for flex as well.

gwwar replied:
Clearing floats tends to be a pain, so if we're able to use flexbox I usually prefer that technique.

gwwar commented:
Nice work @vindl, before 🚢 ing should we disable the sort for Jetpack sites until we use the comments-tree endpoint? 

It works, but we can run into some funny behavior as we load more comments in. For example let's say I have comments from 1-X00. On initial load I have 5 pages of stuff so my oldest comment is somewhere around 200, instead of 1. (which will change as I load in more stuff).

vindl commented:
@gwwar @Copons I pushed some new updates to the PR:

- switched to using flexbox; verified that rtl still works as expected;
- hidden this behind a feature flag that will enable it in development and wpcalypso;
- this was previously nested behind `bulk-edit` feature flag - I've moved that flag closer to the actual bulk edit button to separate the two features (_note_: this will expose empty `CommentNavigationTab` component in production, but that shouldn't cause any issues);
- added a check so it will be only shown for Jetpack sites that have version 5.3 or higher.

gwwar commented on client/my-sites/comments/comment-navigation/index.jsx:
```diff
@@ -230,7 +267,11 @@ const mapStateToProps = ( state, { commentsPage, siteId } ) => {
+
+	return {
+		visibleComments,
+		isCommentsTreeSupported: ! isJetpackSite( state, siteId ) || isJetpackMinimumVersion( state, siteId, '5.3' ),
```
It's fine to leave this here, but just a note that I don't believe we've updated the client to use comments-tree for Jetpack sites yet.

vindl commented:
@folletto @gwwar @Copons @megsfulton Thank you for reviews and suggestions!
